From 1268876e1a5f384209a4e0ff616426de0c0987e9 Mon Sep 17 00:00:00 2001
From: Dario Binacchi <dario.binacchi@amarulasolutions.com>
Date: Mon, 1 Jun 2026 15:59:40 +0200
Subject: [PATCH 3/4] feat(st): add STM32MP_PSA_FWU_AB_SUPPORT build flag

Document and introduce the STM32MP_PSA_FWU_AB_SUPPORT build flag to
conditionally support A/B redundancy systems on STM32MP platforms.

A fallback mechanism is included in the platform Makefile to
automatically override and disable the flag if the core
PSA_FWU_SUPPORT option is missing.

Change-Id: Ic604c448cce1bd180731a0462fcbc0c81f3904d9
Signed-off-by: Dario Binacchi <dario.binacchi@amarulasolutions.com>
Upstream-Status: Backport [https://git.trustedfirmware.org/plugins/gitiles/TF-A/trusted-firmware-a/+/3f01837eb5c726acbccd75de844e2a551c9af11d]
---
 docs/plat/st/stm32mpus.rst |  4 ++++
 plat/st/common/common.mk   | 13 +++++++++++++
 2 files changed, 17 insertions(+)

diff --git a/docs/plat/st/stm32mpus.rst b/docs/plat/st/stm32mpus.rst
index 7a3ea1547598..e9ebdc0da666 100644
--- a/docs/plat/st/stm32mpus.rst
+++ b/docs/plat/st/stm32mpus.rst
@@ -55,6 +55,10 @@ Other configuration flags:
   | default location (end of the first 128MB) is used when absent
 - | ``STM32MP_EARLY_CONSOLE``: to enable early traces before clock driver is setup.
   | Default: 0 (disabled)
+- | ``STM32MP_PSA_FWU_AB_SUPPORT``: to support PSA FWU A/B system.
+  | It requires ``PSA_FWU_SUPPORT`` to be enabled and is only supported on block
+  | devices with a GPT partition table like SD or eMMC.
+  | Default: 0 (disabled)
 - | ``STM32MP_RECONFIGURE_CONSOLE``: to re-configure crash console (especially after BL2).
   | Default: 0 (disabled)
 - | ``STM32MP_UART_BAUDRATE``: to select UART baud rate.
diff --git a/plat/st/common/common.mk b/plat/st/common/common.mk
index e51a03072b67..0b36391288c9 100644
--- a/plat/st/common/common.mk
+++ b/plat/st/common/common.mk
@@ -52,6 +52,13 @@ STM32MP_EMMC_BOOT		?=	0
 STM32MP_UART_PROGRAMMER		?=	0
 STM32MP_USB_PROGRAMMER		?=	0
 
+STM32MP_PSA_FWU_AB_SUPPORT	?=	0
+
+ifeq ($(STM32MP_PSA_FWU_AB_SUPPORT)-$(PSA_FWU_SUPPORT),1-0)
+$(info Disable STM32MP_PSA_FWU_AB_SUPPORT flag. PSA_FWU_SUPPORT is disabled )
+override STM32MP_PSA_FWU_AB_SUPPORT	:=	0
+endif
+
 ifneq ($(filter 1,${STM32MP_UART_PROGRAMMER} ${STM32MP_USB_PROGRAMMER}),)
 ifeq (${PSA_FWU_SUPPORT},1)
 $(info Disable PSA_FWU_SUPPORT flag for serial device)
@@ -74,7 +81,11 @@ endif
 ifeq (${PSA_FWU_SUPPORT},1)
 # Number of banks of updatable firmware
 NR_OF_FW_BANKS			:=	2
+ifeq (${STM32MP_PSA_FWU_AB_SUPPORT},1)
+NR_OF_IMAGES_IN_FW_BANK		:=	3
+else
 NR_OF_IMAGES_IN_FW_BANK		:=	1
+endif
 JSON_METADATA			?=	plat/st/common/default_metadata.json
 
 $(eval $(call GEN_METADATA,${JSON_METADATA},${BUILD_PLAT}/metadata.bin))
@@ -131,6 +142,7 @@ $(eval $(call assert_booleans,\
 		STM32MP_EMMC \
 		STM32MP_EMMC_BOOT \
 		STM32MP_HYPERFLASH \
+		STM32MP_PSA_FWU_AB_SUPPORT \
 		STM32MP_RAW_NAND \
 		STM32MP_RECONFIGURE_CONSOLE \
 		STM32MP_SDMMC \
@@ -154,6 +166,7 @@ $(eval $(call add_defines,\
 		STM32MP_EMMC \
 		STM32MP_EMMC_BOOT \
 		STM32MP_HYPERFLASH \
+		STM32MP_PSA_FWU_AB_SUPPORT \
 		STM32MP_RAW_NAND \
 		STM32MP_RECONFIGURE_CONSOLE \
 		STM32MP_SDMMC \
-- 
2.43.0

